libV2/client/index.js

```diff
@@ -0,0 +1,248 @@
+                })
+                .on('error', reject);
+        });
+    }
+
+    async _drainRetryCachePreflight() {
+        try {
+            const resp = await needle(
+                'get',
+                `http://${this._host}:${this._port}/_/healthcheck`,
+            );
+            return resp.statusCode === 200;
+        } catch (error) {
+            this._logger
+                .debug('drain preflight request failed', { error });
+            return false;
+        }
+    }
+
+    async _attemptDrain() {
+        this._drainTimer = null;
+        if (await this._drainRetryCachePreflight()) {
+            let empty = false;
+
+            try {
+                empty = await this._drainRetryCache();
+            } catch (error) {
+                this._logger
+                    .error('Error while draining cache', { error });
+            }
+
+            if (!empty) {
+                await this._scheduleDrain();
+            }
+        }
+    }
+
+    async _scheduleDrain() {
+        if (this._drainCanSchedule && !this._drainTimer) {
+            this._drainTimer = setTimeout(this._attemptDrain.bind(this), this._drainDelay);
+        }
+    }
+
+    async _disableDrain() {
+        this._drainCanSchedule = false;
+        if (this._drainTimer) {
+            clearTimeout(this._drainTimer);
+            this._drainTimer = null;
+        }
+    }
+
+    _emitMetricLogLine(metric, extra) {
+        this._logger.info('utapi metric recovery log', {
+            event: metric.getValue(),
+            utapiRecovery: true,
+            ...(extra || {}),
+        });
+    }
+
+    async _flushRetryCacheToLogs() {
+        const toRemove = [];
+
+        return new Promise((resolve, reject) => {
+            this._retryCache.createReadStream()
+                .on('data', entry => {
+                    this._emitMetricLogLine(entry.value);
+                    toRemove.push(entry.key);
+                })
+                .on('end', () => {
+                    this._retryCache.batch(toRemove.map(i => ({ type: 'del', key: i })));
+                    resolve();
+                })
+                .on('error', err => reject(err));
+        });
+    }
+
+    async _pushMetric(data) {
+        const metric = data instanceof UtapiMetric
+            ? data
+            : new UtapiMetric(data);
+
+        // Assign a uuid if one isn't passed
+        if (!metric.uuid) {
+            metric.uuid = uuid.v4();
+        }
+
+        // Assign a timestamp if one isn't passed
+        if (!metric.timestamp) {
+            metric.timestamp = new Date().getTime();
+        }
+
+        try {
+            await this._pushToUtapi([metric]);
+        } catch (error) {
+            this._logger.error('unable to push metric, adding to retry cache', { error });
+            if (!await this._addToRetryCache(metric)) {
+                throw new Error('unable to store metric');
+            }
+        }
+    }
+
+    pushMetric(data, cb) {
```

Comment:
I am not sure I understand why cb is here. It seems to be passed in to the _pushMetric method but that doesnt accept a callback.

Reply:
This is to keep compatibility with the existing Utapi client which accepts a callback. If callback is provided `callbackify` is used to convert the promise return by `_pushMetric` to callback style otherwise a Promise is returned

Reply:
ok that makes sense then. Is that the same reason for the other callbacks sprinkled around?

Reply:
The callbacks in `Chunker` and `Uploader` are inherited. The test use whatever I thought would be the most simple, though I did try to lean more toward testing `pushMetric` with callbacks as that's what cloudserver would currently be using

Reply:
Gotcha I see